keymap.c

```diff
@@ -1298,6 +1298,173 @@ enum CommandResult mutt_parse_bind(struct Buffer *buf, struct Buffer *s,
+
+  first = last = NULL;
+  cur = *map;
+
+  while (cur != NULL)
+  {
+    next = cur->next;
+    if (((mode & MUTT_UNBIND) && (cur->macro == NULL)) ||
+        ((mode & MUTT_UNMACRO) && (cur->macro != NULL)))
+    {
+      FREE(&cur->macro);
+      FREE(&cur->keys);
+      FREE(&cur->desc);
+      FREE(&cur);
+    }
+    else if (first == NULL)
+    {
+      first = last = cur;
+    }
+    else if (last != NULL)
+    {
+      last->next = cur;
+      last = cur;
+    }
+    else
+    {
+      last = cur;
+    }
+    cur = next;
+  }
+
+  if (last != NULL)
+    last->next = NULL;
+
+  *map = first;
+}
+
+/**
+ * mutt_parse_unbind - Parse the 'unbind' command - Implements ::command_t
+ * @param buf
+ * @param s
+ * @param data
+ * @param err
+ *
+ * @return r
+ *
+ * Command unbinds specific binging in menu-name, in all menu-names
+ * or all bindings in all menu-names.
+ * unbind <menu-name[,…]|*> [<key_sequence>]
+ */
+enum CommandResult mutt_parse_unbind(struct Buffer *buf, struct Buffer *s,
+                                     unsigned long data, struct Buffer *err)
+{
+  int menu[sizeof(Menus) / sizeof(struct Mapping) - 1], nummenus;
+  enum CommandResult r = MUTT_CMD_SUCCESS;
+  bool all_menus = false, all_keys = false;
+  char *key = NULL;
+  char *cmd = "unbind";
+  if ((data & MUTT_UNMACRO) == MUTT_UNMACRO)
+    cmd = "unmacro";
+
+  mutt_extract_token(buf, s, 0);
+  if (mutt_str_strcmp(buf->data, "*") == 0) //when menu-names is * (all)
+    all_menus = true;
+  else // when menu-names is comma separated list
+    parse_menu(menu, buf->data, mutt_array_size(menu), &nummenus, err, cmd);
+
+  if (MoreArgs(s)) // second arg - key
+  {
+    mutt_extract_token(buf, s, 0);
+    key = buf->data;
+  }
+  else
+    all_keys = true;
+
+  if ((all_menus && mutt_str_strcmp(key, "q") == 0) ||
+      (all_menus && mutt_str_strcmp(key, ":") == 0))
+  {
+    mutt_buffer_printf(err, _("%s: cannot unbind '%s' from all menus"), cmd, key);
+    return MUTT_CMD_ERROR;
+  }
+
+  if (MoreArgs(s))
+  {
+    mutt_buffer_printf(err, _("%s: too many arguments"), cmd);
+    return MUTT_CMD_ERROR;
+  }
+  else // Here comes the logic, Mr. Tuvok.
+  {
+    if (all_keys)
+    {
+      int menus = all_menus ? MENU_MAX : nummenus;
+
+      for (int i = 0; i < menus; ++i)
+      {
+        int map = all_menus ? i : menu[i];
+        km_unbind_all(&Keymaps[map],data);
+        if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
+        {
+          km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
+          km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
+          km_bindkey("q", MENU_GENERIC, OP_EXIT);
+          km_bindkey("q", MENU_PAGER, OP_EXIT);
+          km_bindkey("?", MENU_GENERIC, OP_HELP);
+        }
+      }
+    }
+    else
+    {
+      int menus = all_menus ? MENU_MAX : nummenus;
+      for (int i = 0; i < menus; ++i) {
```

Comment:
After making the changes, both parts of the `if` have the same basic format. We can change extract the `for` loop so the conditional is within in and have something like:

```C
int menus = all_menus ? MENU_MAX : nummenus;

for (int i = 0; i < menus; ++i) 
{
    int map = all_menus ? i : menu[i];

    if (all_keys) 
    {
        km_unbind_all(&Keymaps[map],data);
        if ((data & MUTT_UNBIND) == MUTT_UNBIND && i == MENU_GENERIC)
        {
          km_bindkey("<return>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
          km_bindkey("<enter>", MENU_GENERIC, OP_GENERIC_SELECT_ENTRY);
          km_bindkey(":", MENU_GENERIC, OP_ENTER_COMMAND);
          km_bindkey("q", MENU_GENERIC, OP_EXIT);
          km_bindkey("q", MENU_PAGER, OP_EXIT);
          km_bindkey("?", MENU_GENERIC, OP_HELP);
        }
    }
    else
    {
        km_bindkey(key, map, OP_NULL);
    }
}
```

Also, since `if (MoreArgs(s))` has a `return` in it, we can remove the `else` and have the `for` loop directly after. Reduces some indentation and number of braces.